core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -0,0 +1,658 @@
+ * page.
+ *
+ * Because the application cache will detach and then re-attach the UI, when the caller
+ * repeats that request, it will now pick up the newly-updated web application.
+ *
+ * This does require the caller to handle 302 requests. Because of the ambiguity
+ * in how POST and PUT operations are responded to (that is, should a 307 be
+ * processed directly), the filter <i>does not</i> filter those requests.
+ * As the current web UIs are read-only, this is not an issue. If it were ever to
+ * support more HTTP verbs, then some support may be required. Perhaps, rather
+ * than sending a redirect, simply updating the value so that the <i>next</i>
+ * request will pick it up.
+ *
+ * Implementation note: there's some abuse of a shared global entry here because
+ * the configuration data passed to the servlet is just a string:string map.
+ */
+private[history] class ApplicationCacheCheckFilter() extends Filter with Logging {
+
+  import ApplicationCacheCheckFilterRelay._
+  var appId: String = _
+  var attemptId: Option[String] = _
+
+  /**
+   * Bind the app and attempt ID, throwing an exception if no application ID was provided.
+   * @param filterConfig configuration
+   */
+  override def init(filterConfig: FilterConfig): Unit = {
+
+    appId = Option(filterConfig.getInitParameter(APP_ID))
+      .getOrElse(throw new ServletException(s"Missing Parameter $APP_ID"))
+    attemptId = Option(filterConfig.getInitParameter(ATTEMPT_ID))
+    logDebug(s"initializing filter $this")
+  }
+
+  /**
+   * Filter the request.
+   * Either the caller is given a 302 redirect to the current URL, or the
+   * request is passed on to the SparkUI servlets.
+   *
+   * @param request HttpServletRequest
+   * @param response HttpServletResponse
+   * @param chain the rest of the request chain
+   */
+  override def doFilter(
+      request: ServletRequest,
+      response: ServletResponse,
+      chain: FilterChain): Unit = {
+
+    // nobody has ever implemented any other kind of servlet, yet
+    // this check is universal, just in case someone does exactly
+    // that on your classpath
+    if (!(request.isInstanceOf[HttpServletRequest])) {
+      throw new ServletException("This filter only works for HTTP/HTTPS")
+    }
+    val httpRequest = request.asInstanceOf[HttpServletRequest]
+    val httpResponse = response.asInstanceOf[HttpServletResponse]
+    val requestURI = httpRequest.getRequestURI
+    val operation = httpRequest.getMethod
+
+    // if the request is for an attempt, check to see if it is in need of delete/refresh
+    // and have the cache update the UI if so
+    if (operation=="HEAD" || operation=="GET"
+        && checkForUpdates(requestURI, appId, attemptId)) {
+      // send a redirect back to the same location. This will be routed
+      // to the *new* UI
+      logInfo(s"Application Attempt $appId/$attemptId updated; refreshing")
+      val redirectUrl = httpResponse.encodeRedirectURL(requestURI)
```

Comment:
ah, now I see how to recreate it -- open a browser to the job listing page of the history server.  Go run another job in your spark-shell, so the app has updated, but do *not* refresh the page in browser.  Wait long enough that the history server will call the app stale.  Now click on one of the existing links.  The redirect is broken.

btw, since this took me a bit to figure out, I thought I'd share how I ran things locally if you want to as well in case it helps:

1) run the history server from within sbt using the sbt-revolver plugin: `re-start --- -Dspark.history.fs.logDirectory=<path> -Dspark.history.fs.update.interval=5s -Dspark.history.cache.window=5s`
2) select the HistoryServer option from the list
3) starta  spark-shell with `bin/spark-shell --conf "spark.eventLog.dir=<path>" --conf "spark.eventLog.enabled=true"`

this way you can start & stop the history server quickly as you make code changes.